pkg/services/store/entity/sqlstash/sql_storage_server.go

```diff
@@ -1127,31 +1082,44 @@ func (s *sqlEntityServer) List(ctx context.Context, r *entity.EntityListRequest)
 
 	fields := s.getReadFields(r)
 
-	entityQuery := selectQuery{
-		dialect:  s.dialect,
-		fields:   fields,
-		from:     "entity", // the table
-		args:     []any{},
-		limit:    r.Limit,
-		offset:   0,
-		oneExtra: true, // request one more than the limit (and show next token if it exists)
-	}
+	// main query we will use to retrieve entities
+	entityQuery := NewSelectQuery(s.dialect, entityTable)
+	entityQuery.AddFields(fields...)
+	entityQuery.SetLimit(r.Limit)
+	entityQuery.SetOneExtra()
+
+	// query to retrieve the max resource version and entity count
+	rvMaxQuery := NewSelectQuery(s.dialect, entityTable)
```

Comment:
_nit. This is maybe just a todo given it will only start to be a problem when we add some throughput through the system._

One challenge: There is no strict guarantee about ordering: two concurrent statements may have their commit time in one order, but their RV in the opposite order.

So, when we find out the highest resource_version, it's likely to have uncommited events with RV < max(RV) from all the uncommited concurrent transactions. The List operation will not reflect those changes (given the transations are not commited), but also client will never see those events given the RV is lower than the max RV. 

The only way I know to fix this is to **start** the first statement in the transaction by doing a [SELECT ... FOR SHARE](https://dev.mysql.com/doc/refman/8.0/en/select.html) (i.e `SELECT count(*) where <conditions> FOR SHARE`). This will ensure we wait for all the transactions in progress to commit before we can move on. Sadly this will also provent concurrent updates to the very same objects while the transaction is running.

Reply:
Yeah, we're going to have to think more carefully about locking, I was trying to avoid needed a global lock but we may not be able to avoid it.